File: 2023/day2/part2.js
const fs = require('fs');
const readline = require('readline');

if(process.argv.length<3) {
  console.log("Pass path to data file as argument");
  return 1;
}
var rd = readline.createInterface({
  input: fs.createReadStream(argv[2]),
});
let acc = 0
rd.on('line',
  line => {
    const labeled = line.split(": ")
    const number = parseInt(labeled[0].split(" ")[1])
    const games = labeled[1].split("; ")
    const takes = games.map(x=>x.split(", ").map(y=>y.split(" "))).flat()
    .map(x=>[parseInt(x[0]),x[1]])
    .reduce((acc,tuple)=>(
      (tuple[1][0]=='r')
      ?(tuple[0]>acc[0])?[tuple[0],acc[1],acc[2]]:acc
      :(tuple[1][0]=='g')
        ?(tuple[0]>acc[1])?[acc[0],tuple[0],acc[2]]:acc
        :(tuple[0]>acc[2])?[acc[0],acc[1],tuple[0]]:acc),[0,0,0])
    const power = takes[0]*takes[1]*takes[2]
    acc += power
  })
rd.on('close',
  ()=>console.log(acc))

File: 2023/day2/scala/src/main/scala/Part2.scala
package main.scala
import scala.io.Source._
import scala.math.max

object Part2 {
  def main(args: Array[String]): Unit = {
    if(args.length<1) {
      println("Pass path to data file as argument")
      return;
    }
    val lines = fromFile(args(0)).getLines
      .map(line=>line.split(": "))
      .map(labeled=>(labeled(0).split(" ")(1).toInt,labeled(1)))
      .map(numbered=>(numbered(0),
        numbered(1).toString().split("; ").toList
          .map(game=>game.split(", ").toList
          .map(color=>{
            val s = color.split(" ")
            (s(0).toInt,s(1)(0))
          })).flatten
            .foldLeft((0,0,0))((acc,x)=>{
              x(1) match
              case 'r' => (max(acc(0),x(0)),acc(1),acc(2))
              case 'g' => (acc(0),max(acc(1),x(0)),acc(2))
              case 'b' => (acc(0),acc(1),max(acc(2),x(0)))
            })))
              .foldLeft(0)((acc,numbered)=>
                  acc+numbered(1)(0)*numbered(1)(1)*numbered(1)(2))
    println(lines)
  }
}
